src/App.tsx

```diff
@@ -1,26 +1,241 @@
+
+    try {
+      await client.delete(`/todos/${todoId}`);
+      setTodos(prev => prev.filter(t => t.id !== todoId));
+    } catch {
+      setErrorMessage('Unable to delete a todo');
+    } finally {
+      setDeletingIds(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  const handleToggle = async (todo: Todo) => {
+    if (updatingIds.includes(todo.id)) {
+      return;
+    }
+
+    setUpdatingIds(prev => [...prev, todo.id]);
+
+    try {
+      const updated = await updateTodo({ ...todo, completed: !todo.completed });
+
+      setTodos(prev => prev.map(t => (t.id === todo.id ? updated : t)));
+    } catch {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingIds(prev => prev.filter(id => id !== todo.id));
+    }
+  };
+
+  const handleRename = async (todo: Todo, newTitle: string) => {
+    if (updatingIds.includes(todo.id)) {
+      return;
+    }
+
+    setUpdatingIds(prev => [...prev, todo.id]);
+
+    try {
+      const updated = await updateTodo({ ...todo, title: newTitle });
+
+      setTodos(prev => prev.map(t => (t.id === todo.id ? updated : t)));
+    } catch {
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setUpdatingIds(prev => prev.filter(id => id !== todo.id));
+    }
+  };
+
+  const handleToggleAll = async (completed: boolean) => {
+    const toUpdate = todos.filter(t => t.completed !== completed);
+
+    if (!toUpdate.length) {
+      return;
+    }
+
+    const ids = toUpdate.map(t => t.id);
+
+    setUpdatingIds(prev => [...prev, ...ids]);
+
+    const promises = toUpdate.map(t =>
+      updateTodo({ ...t, completed })
+        .then(() => ({ success: true, id: t.id }))
+        .catch(() => ({ success: false, id: t.id })),
+    );
+
+    const results = await Promise.all(promises);
+    const okIds = results.filter(r => r.success).map(r => r.id);
+
+    if (okIds.length) {
+      setTodos(prev =>
+        prev.map(t => (okIds.includes(t.id) ? { ...t, completed } : t)),
+      );
+    }
+
+    const hasError = results.some(r => !r.success);
+
+    if (hasError) {
+      setErrorMessage('Unable to update a todo');
+    }
+
+    setUpdatingIds(prev => prev.filter(id => !ids.includes(id)));
+  };
+
+  const handleClearCompleted = async () => {
+    const completed = todos.filter(t => t.completed);
+
+    setDeletingIds(completed.map(t => t.id));
+
+    const promises = completed.map(t =>
+      client
+        .delete(`/todos/${t.id}`)
+        .then(() => ({ ok: true, id: t.id }))
+        .catch(() => ({ ok: false, id: t.id })),
+    );
+
+    const results = await Promise.all(promises);
+    const okIds = results.filter(r => r.ok).map(r => r.id);
+
+    if (okIds.length) {
+      setTodos(prev => prev.filter(t => !okIds.includes(t.id)));
+    }
+
+    const hasError = results.some(r => !r.ok);
+
+    if (hasError) {
+      setErrorMessage('Unable to delete a todo');
+    }
+
+    setDeletingIds([]);
+  };
+
+  const getFilteredTodos = (): Todo[] => {
```

Comment:
You can move this function in another file, making it clean (pure function) and divide app component